File: ui/src/lib/automations.ts
import { request } from "./transport";

/** Scheduled automations (`/api/automations…`, docs/API_CONTRACT_0.28.md
 * and docs/AUTOMATIONS.md). Times are Unix seconds. */

export type Schedule =
  | { kind: "hourly"; minute: number }
  | { kind: "daily"; time: string }
  | { kind: "weekdays"; time: string }
  | { kind: "weekly"; day: number; time: string }
  | { kind: "cron"; expr: string };

export type AutomationOptions = {
  checkout: "worktree" | "main";
  permission: "project" | "read_only";
  on_approval: "stop" | "wait";
  max_runtime_minutes: number;
  catch_up_minutes: number;
  notify: boolean;
};

export type AutomationMode = "code" | "plan" | "ask";

export type AutomationDraft = {
  name: string;
  prompt: string;
  /** A picker id; "" runs on the project's model. */
  model: string;
  mode: AutomationMode;
  schedule: Schedule;
  timezone: "local" | "utc";
  options: AutomationOptions;
};

export type RunStatus =
  | "running"
  | "completed"
  | "failed"
  | "cancelled"
  | "timed_out"
  | "needs_approval"
  | "interrupted"
  | "missed"
  | "skipped";

export type AutomationRun = {
  id: string;
  automation_id: string;
  status: RunStatus | string;
  trigger: "schedule" | "catch_up" | "manual" | string;
  scheduled_for?: number | null;
  started_at: number;
  finished_at?: number | null;
  duration?: number | null;
  session_id?: string | null;
  job_id?: string | null;
  summary?: string;
  detail?: string;
  usage?: {
    total_tokens?: number;
    cost_usd?: number | null;
    cost_estimated?: boolean;
  } | null;
  worktree?: {
    id: string;
    path: string;
    branch: string;
    removed?: boolean;
  } | null;
  missed?: number;
};

export type Automation = AutomationDraft & {
  id: string;
  workspace: string;
  paused: boolean;
  next_run_at: number | null;
  created_at: number;
  updated_at: number;
  description: string;
  running_run: string | null;
  last_run: AutomationRun | null;
  runs?: AutomationRun[];
};

export type AutomationList = {
  workspace: string;
  automations: Automation[];
  /** False when no desktop or `shadowcode serve` runs the schedules. */
  scheduler: boolean;
  now: number;
};

export type SchedulePreview =
  | { ok: true; description: string; next: number[]; now: number }
  | { ok: false; error: string };

export const automationsApi = {
  list: () => request<AutomationList>("/api/automations"),
  get: (id: string) => request<Automation>(`/api/automations/${id}`),
  create: (draft: AutomationDraft) =>
    request<Automation>("/api/automations", "POST", draft),
  update: (id: string, draft: AutomationDraft) =>
    request<Automation>(`/api/automations/${id}`, "POST", draft),
  remove: (id: string) =>
    request<{ ok: boolean }>(`/api/automations/${id}`, "DELETE"),
  pause: (id: string) =>
    request<Automation>(`/api/automations/${id}/pause`, "POST", {}),
  resume: (id: string) =>
    request<Automation>(`/api/automations/${id}/resume`, "POST", {}),
  runNow: (id: string) =>
    request<AutomationRun>(`/api/automations/${id}/run`, "POST", {}),
  stop: (id: string) =>
    request<AutomationRun>(`/api/automations/${id}/stop`, "POST", {}),
  preview: (schedule: Schedule, timezone: "local" | "utc") =>
    request<SchedulePreview>("/api/automations/preview", "POST", {
      schedule,
      timezone,
    }),
};

export const DAYS = [
  "Sunday",
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
];

export function defaultDraft(repo: boolean): AutomationDraft {
  return {
    name: "",
    prompt: "",
    model: "",
    mode: "code",
    schedule: { kind: "weekdays", time: "09:00" },
    timezone: "local",
    options: {
      checkout: repo ? "worktree" : "main",
      permission: "project",
      on_approval: "stop",
      max_runtime_minutes: 60,
      catch_up_minutes: 120,
      notify: true,
    },
  };
}

export function draftOf(a: Automation): AutomationDraft {
  return {
    name: a.name,
    prompt: a.prompt,
    model: a.model,
    mode: a.mode,
    schedule: a.schedule,
    timezone: a.timezone,
    options: a.options,
  };
}

/** Switch schedule kind, keeping the time the user already chose. */
export function withKind(schedule: Schedule, kind: Schedule["kind"]): Schedule {
  const time = "time" in schedule ? schedule.time : "09:00";
  switch (kind) {
    case "hourly":
      return { kind, minute: 0 };
    case "daily":
    case "weekdays":
      return { kind, time };
    case "weekly":
      return { kind, day: 1, time };
    case "cron":
      return { kind, expr: "0 9 * * 1-5" };
  }
}

const pad = (n: number) => String(n).padStart(2, "0");

/** "in 5 min", "Today 09:00", "Tomorrow 09:00", "Mon 28 Sep 09:00" —
 * in the viewer's local time. */
export function formatWhen(ts: number, now: number): string {
  const at = new Date(ts * 1000);
  const current = new Date(now * 1000);
  const minutes = Math.round((ts - now) / 60);
  if (minutes >= 0 && minutes < 60)
    return minutes <= 1 ? "in a minute" : `in ${minutes} min`;
  const clock = `${pad(at.getHours())}:${pad(at.getMinutes())}`;
  const day = (d: Date) =>
    new Date(d.getFullYear(), d.getMonth(), d.getDate()).getTime();
  const days = Math.round((day(at) - day(current)) / 86_400_000);
  if (days === 0) return `Today ${clock}`;
  if (days === 1) return `Tomorrow ${clock}`;
  if (days === -1) return `Yesterday ${clock}`;
  const month = at.toLocaleString("en", { month: "short" });
  const weekday = at.toLocaleString("en", { weekday: "short" });
  return `${weekday} ${at.getDate()} ${month} ${clock}`;
}

/** The list's "Next:" line. */
export function nextRunText(a: Automation, now: number): string {
  if (a.running_run) return "Running now";
  if (a.paused) return "Paused";
  if (!a.next_run_at) return "No upcoming time";
  return `Next: ${formatWhen(a.next_run_at, now)}`;
}

export function formatDuration(seconds: number | null | undefined): string {
  if (seconds == null) return "";
  if (seconds < 60) return `${Math.max(1, Math.round(seconds))} s`;
  const minutes = Math.round(seconds / 60);
  if (minutes < 60) return `${minutes} min`;
  return `${Math.floor(minutes / 60)} h ${minutes % 60} min`;
}

export function formatCost(run: AutomationRun): string {
  const usage = run.usage;
  if (!usage) return "";
  if (usage.cost_usd != null)
    return `${usage.cost_estimated ? "about " : ""}$${usage.cost_usd.toFixed(
      usage.cost_usd < 0.01 ? 4 : 2,
    )}`;
  if (usage.total_tokens)
    return `${usage.total_tokens.toLocaleString()} tokens`;
  return "";
}

export const STATUS_TEXT: Record<string, string> = {
  running: "Running",
  completed: "Finished",
  failed: "Failed",
  cancelled: "Stopped",
  timed_out: "Hit its time limit",
  needs_approval: "Stopped for approval",
  interrupted: "Interrupted",
  missed: "Missed",
  skipped: "Skipped",
};

export function runText(run: AutomationRun): string {
  const status = STATUS_TEXT[run.status] || run.status;
  if (run.status === "missed" && (run.missed || 0) > 1)
    return `${status} (${run.missed} times)`;
  return status;
}

export const TRIGGER_TEXT: Record<string, string> = {
  schedule: "On schedule",
  catch_up: "Caught up after start",
  manual: "Run now",
};

File: native/core/tests/automations.rs
//! Scheduled automations end to end against a fake model: runs in their own
//! tagged conversations, history with status/duration/usage, no overlapping
//! runs, catch-up and missed times, fresh worktrees that are removed when
//! nothing changed, approval requests stopping unattended runs, and
//! recovery of runs interrupted by a restart.
mod support;
use serde_json::{json, Value};
use shadowcode_core::{
    config::Config,
    paths::AppPaths,
    service::{Request, Service},
    store::{keys, AutomationRun, Store},
};
use std::{
    fs,
    path::{Path, PathBuf},
    process::Command,
    time::Duration,
};

fn response(text: &str, calls: Value) -> Value {
    json!({"choices":[{"message":{"role":"assistant","content":text,"tool_calls":calls},"finish_reason":if calls.as_array().is_some_and(|a|!a.is_empty()){"tool_calls"}else{"stop"}}],"usage":{"prompt_tokens":20,"completion_tokens":10,"total_tokens":30}})
}
fn tool(name: &str, args: Value) -> Value {
    json!({"id":shadowcode_core::id(),"type":"function","function":{"name":name,"arguments":args.to_string()}})
}
fn git(dir: &Path, args: &[&str]) {
    let result = Command::new("git")
        .args([
            "-c",
            "core.hooksPath=/dev/null",
            "-c",
            "commit.gpgSign=false",
        ])
        .args(args)
        .current_dir(dir)
        .output()
        .unwrap();
    assert!(
        result.status.success(),
        "git {args:?}: {}",
        String::from_utf8_lossy(&result.stderr)
    );
}

struct Fixture {
    _root: tempfile::TempDir,
    project: PathBuf,
    paths: AppPaths,
    service: Service,
}

fn setup(endpoint: &str, approve_shell: bool) -> Fixture {
    let root = tempfile::tempdir().unwrap();
    let project = root.path().join("project");
    fs::create_dir(&project).unwrap();
    git(&project, &["init", "-b", "main"]);
    git(&project, &["config", "user.name", "Automation Test"]);
    git(
        &project,
        &["config", "user.email", "automation@example.invalid"],
    );
    fs::write(project.join("README.md"), "# Demo\n").unwrap();
    git(&project, &["add", "."]);
    git(&project, &["commit", "-m", "Initial commit"]);
    let project = project.canonicalize().unwrap();
    let paths = AppPaths::isolated(&root.path().join("profile")).unwrap();
    Config::patch(&paths, json!({"model":{"provider":"local","endpoint":endpoint,"name":"fixture","context_limit":16384},"trusted_workspaces":[project],"permissions":{"approve_shell":approve_shell},"agent":{"max_steps":8,"retry_attempts":0}})).unwrap();
    let service = Service::open(paths.clone(), Some(project.clone())).unwrap();
    Fixture {
        _root: root,
        project,
        paths,
        service,
    }
}

async fn call(service: &Service, method: &str, path: &str, body: Value) -> anyhow::Result<Value> {
    service
        .dispatch(Request {
            method: method.into(),
            path: path.into(),
            body,
        })
        .await
}

fn automation(name: &str, checkout: &str) -> Value {
    json!({
        "name": name,
        "prompt": "Look at the project and report.",
        "mode": "code",
        "schedule": {"kind": "hourly", "minute": 0},
        "timezone": "utc",
        "options": {"checkout": checkout, "max_runtime_minutes": 5, "catch_up_minutes": 60},
    })
}

async fn finished(fixture: &Fixture, id: &str) -> AutomationRun {
    tokio::time::timeout(
        Duration::from_secs(20),
        fixture.service.engine.wait_automation(id),
    )
    .await
    .expect("automation run finished")
    .unwrap()
    .expect("a history row")
}

#[tokio::test]
async fn runs_are_tagged_recorded_and_never_overlap() {
    let server = support::server(|_, _| {
        (
            response("Checked the project; all good.", json!([])),
            Duration::from_millis(600),
        )
    })
    .await;
    let fixture = setup(&server.endpoint, false);
    let service = &fixture.service;
    let created = call(
        service,
        "POST",
        "/api/automations",
        automation("Hourly check", "main"),
    )
    .await
    .unwrap();
    let id = created["id"].as_str().unwrap().to_owned();
    assert_eq!(created["description"], "Every hour at :00 UTC");
    let next = created["next_run_at"].as_f64().unwrap();
    assert!(next > shadowcode_core::now() && next <= shadowcode_core::now() + 3600.0);
    assert_eq!(created["options"]["on_approval"], "stop", "safe default");

    let run = call(
        service,
        "POST",
        &format!("/api/automations/{id}/run"),
        json!({}),
    )
    .await
    .unwrap();
    assert_eq!(run["status"], "running");
    assert_eq!(run["trigger"], "manual");
    // A second "Run now" and a schedule tick during the run both refuse to
    // start another one.
    let again = call(
        service,
        "POST",
        &format!("/api/automations/{id}/run"),
        json!({}),
    )
    .await
    .unwrap_err()
    .to_string();
    assert!(again.contains("already running"), "{again}");
    let rows = service.engine.automation_tick(next + 1.0).await.unwrap();
    assert_eq!(rows.len(), 1);
    assert_eq!(rows[0].status, "skipped");
    let listed = call(service, "GET", "/api/automations", Value::Null)
        .await
        .unwrap();
    assert_eq!(
        listed["automations"][0]["running_run"], run["id"],
        "the list shows the active run"
    );
    assert!(listed["automations"][0]["next_run_at"].as_f64().unwrap() > next);

    let done = finished(&fixture, &id).await;
    assert_eq!(done.id, run["id"].as_str().unwrap());
    assert_eq!(done.status, "completed", "{done:?}");
    assert!(done.summary.contains("all good"), "{}", done.summary);
    assert!(done.duration().unwrap() >= 0.5);
    assert_eq!(done.usage.as_ref().unwrap()["total_tokens"], 30);
    // The model received the automation's prompt as the task.
    assert!(server.requests.lock().unwrap().iter().any(|r| r["messages"]
        .to_string()
        .contains("Look at the project and report.")));
    // Its own conversation, tagged with the automation, in the project.
    let store = service.engine.store();
    let sid = done.session_id.clone().unwrap();
    assert_eq!(
        store.session_meta(&sid, keys::AUTOMATION_ID).unwrap(),
        Some(id.clone())
    );
    assert_eq!(
        store.session_meta(&sid, keys::AUTOMATION_RUN).unwrap(),
        Some(done.id.clone())
    );
    let session = store.session(&sid).unwrap().unwrap();
    assert_eq!(session["workspace"], json!(fixture.project));
    assert_eq!(session["title"], "Hourly check · automation");

    let detail = call(
        service,
        "GET",
        &format!("/api/automations/{id}"),
        Value::Null,
    )
    .await
    .unwrap();
    let runs = detail["runs"].as_array().unwrap();
    let mut statuses: Vec<_> = runs
        .iter()
        .map(|r| r["status"].as_str().unwrap().to_owned())
        .collect();
    statuses.sort();
    assert_eq!(statuses, ["completed", "skipped"]);
    let completed = runs.iter().find(|r| r["status"] == "completed").unwrap();
    assert!(completed["duration"].as_f64().unwrap() > 0.0);
    assert_eq!(detail["running_run"], Value::Null);

    // Pause stops the schedule; resume starts it from now; delete removes it.
    let paused = call(
        service,
        "POST",
        &format!("/api/automations/{id}/pause"),
        json!({}),
    )
    .await
    .unwrap();
    assert_eq!(paused["paused"], true);
    assert_eq!(paused["next_run_at"], Value::Null);
    assert!(service
        .engine
        .automation_tick(next + 7200.0)
        .await
        .unwrap()
        .is_empty());
    let resumed = call(
        service,
        "POST",
        &format!("/api/automations/{id}/resume"),
        json!({}),
    )
    .await
    .unwrap();
    assert_eq!(resumed["paused"], false);
    assert!(resumed["next_run_at"].as_f64().unwrap() > shadowcode_core::now());
    call(
        service,
        "DELETE",
        &format!("/api/automations/{id}"),
        Value::Null,
    )
    .await
    .unwrap();
    assert!(call(service, "GET", "/api/automations", Value::Null)
        .await
        .unwrap()["automations"]
        .as_array()
        .unwrap()
        .is_empty());
}

#[tokio::test]
async fn missed_times_catch_up_within_the_window_and_are_recorded_otherwise() {
    let server = support::server(|_, _| (response("Done.", json!([])), Duration::ZERO)).await;
    let fixture = setup(&server.endpoint, false);
    let service = &fixture.service;
    let store = service.engine.store();
    let recent = call(
        service,
        "POST",
        "/api/automations",
        automation("Recent", "main"),
    )
    .await
    .unwrap();
    let old = call(
        service,
        "POST",
        "/api/automations",
        automation("Old", "main"),
    )
    .await
    .unwrap();
    let paused = call(
        service,
        "POST",
        "/api/automations",
        automation("Paused", "main"),
    )
    .await
    .unwrap();
    let (recent, old, paused) = (
        recent["id"].as_str().unwrap(),
        old["id"].as_str().unwrap(),
        paused["id"].as_str().unwrap(),
    );
    call(
        service,
        "POST",
        &format!("/api/automations/{paused}/pause"),
        json!({}),
    )
    .await
    .unwrap();
    // As if ShadowCode had been closed: one time 30 minutes ago (inside the
    // 60 minute window), one five hours ago (outside it).
    let now = 1_900_000_000.0; // a fixed instant: 2030-03-17T17:46:40Z
    store
        .set_automation_next_run(recent, Some(now - 1800.0))
        .unwrap();
    store
        .set_automation_next_run(old, Some(now - 5.0 * 3600.0 - 60.0))
        .unwrap();
    let rows = service.engine.automation_tick(now).await.unwrap();
    assert_eq!(rows.len(), 2, "{rows:?}");
    let caught = rows.iter().find(|r| r.automation_id == recent).unwrap();
    assert_eq!(caught.trigger, "catch_up");
    assert_eq!(caught.status, "running");
    let missed = rows.iter().find(|r| r.automation_id == old).unwrap();
    assert_eq!(missed.status, "missed");
    assert_eq!(missed.missed, 6, "hourly times from 5h01m ago until now");
    assert!(missed.detail.contains("catch-up window"));
    // Both move to their next time after `now`; nothing fires twice.
    for id in [recent, old] {
        let next = store.automation(id).unwrap().next_run_at.unwrap();
        assert!(next > now && next <= now + 3600.0);
    }
    assert!(service
        .engine
        .automation_tick(now)
        .await
        .unwrap()
        .is_empty());
    assert_eq!(finished(&fixture, recent).await.status, "completed");
    assert!(store.automation_runs(paused, 10).unwrap().is_empty());
}

#[tokio::test]
async fn worktree_runs_remove_an_untouched_checkout_and_keep_one_with_changes() {
    let server = support::server(|_, body| {
        let messages = body["messages"].as_array().unwrap();
        let answered = messages.iter().any(|m| m["role"] == "tool");
        let value = if body["messages"].to_string().contains("Write the report") && !answered {
            response(
                "Writing",
                json!([tool(
                    "write_file",
                    json!({"path":"report.txt","content":"ok\n","expected_hash":"missing"})
                )]),
            )
        } else {
            response("Finished.", json!([]))
        };
        (value, Duration::ZERO)
    })
    .await;
    let fixture = setup(&server.endpoint, false);
    let service = &fixture.service;
    let store = service.engine.store();

    let quiet = call(
        service,
        "POST",
        "/api/automations",
        automation("Quiet", "worktree"),
    )
    .await
    .unwrap();
    let quiet = quiet["id"].as_str().unwrap();
    call(
        service,
        "POST",
        &format!("/api/automations/{quiet}/run"),
        json!({}),
    )
    .await
    .unwrap();
    let run = finished(&fixture, quiet).await;
    assert_eq!(run.status, "completed", "{run:?}");
    let tree = run.worktree.clone().unwrap();
    assert_eq!(tree["removed"], true, "{run:?}");
    assert!(!Path::new(tree["path"].as_str().unwrap()).exists());
    assert!(run.detail.contains("temporary worktree was removed"));
    // Its conversation now points at the project, and the checkout is no
    // longer trusted.
    let session = store
        .session(run.session_id.as_deref().unwrap())
        .unwrap()
        .unwrap();
    assert_eq!(session["workspace"], json!(fixture.project));
    let cfg = Config::load(&fixture.paths, None).unwrap();
    assert!(!cfg.is_trusted(Path::new(tree["path"].as_str().unwrap())));

    let mut busy = automation("Writer", "worktree");
    busy["prompt"] = json!("Write the report file.");
    let busy = call(service, "POST", "/api/automations", busy)
        .await
        .unwrap();
    let busy = busy["id"].as_str().unwrap();
    call(
        service,
        "POST",
        &format!("/api/automations/{busy}/run"),
        json!({}),
    )
    .await
    .unwrap();
    let run = finished(&fixture, busy).await;
    assert_eq!(run.status, "completed", "{run:?}");
    let tree = run.worktree.clone().unwrap();
    assert_ne!(tree["removed"], true);
    let checkout = PathBuf::from(tree["path"].as_str().unwrap());
    assert_eq!(
        fs::read_to_string(checkout.join("report.txt")).unwrap(),
        "ok\n"
    );
    assert!(
        !fixture.project.join("report.txt").exists(),
        "the project itself is untouched"
    );
    assert!(
        run.detail.contains(tree["branch"].as_str().unwrap()),
        "{}",
        run.detail
    );
    // Opening that conversation never makes the worktree a project.
    let sid = run.session_id.clone().unwrap();
    call(
        service,
        "POST",
        &format!("/api/sessions/{sid}/activate"),
        json!({}),
    )
    .await
    .unwrap();
    assert_eq!(service.workspace().unwrap(), checkout);
    let projects = call(service, "GET", "/api/projects", Value::Null)
        .await
        .unwrap();
    assert!(
        !projects.to_string().contains(checkout.to_str().unwrap()),
        "{projects}"
    );
}

#[tokio::test]
async fn approval_requests_stop_an_unattended_run() {
    let server = support::server(|_, _| {
        (
            response(
                "Running tests",
                json!([tool("exec", json!({"command":"echo hi"}))]),
            ),
            Duration::ZERO,
        )
    })
    .await;
    let fixture = setup(&server.endpoint, true);
    let service = &fixture.service;
    let created = call(
        service,
        "POST",
        "/api/automations",
        automation("Tests", "main"),
    )
    .await
    .unwrap();
    let id = created["id"].as_str().unwrap();
    call(
        service,
        "POST",
        &format!("/api/automations/{id}/run"),
        json!({}),
    )
    .await
    .unwrap();
    let run = finished(&fixture, id).await;
    assert_eq!(run.status, "needs_approval", "{run:?}");
    assert!(run.detail.contains("echo hi"), "{}", run.detail);
    let job = service
        .engine
        .job(run.job_id.as_deref().unwrap())
        .unwrap()
        .unwrap();
    assert_eq!(job.status, "cancelled");
}

#[tokio::test]
async fn invalid_automations_and_untrusted_projects_are_refused() {
    let server = support::server(|_, _| (response("x", json!([])), Duration::ZERO)).await;
    let fixture = setup(&server.endpoint, false);
    let service = &fixture.service;
    for (patch, expected) in [
        (json!({"name": ""}), "name"),
        (json!({"mode": "yolo"}), "Mode"),
        (
            json!({"schedule": {"kind": "cron", "expr": "61 * * * *"}}),
            "minute",
        ),
        (
            json!({"schedule": {"kind": "cron", "expr": "0 0 30 2 *"}}),
            "never runs",
        ),
        (json!({"options": {"max_runtime_minutes": 0}}), "time limit"),
    ] {
        let mut body = automation("Bad", "main");
        for (key, value) in patch.as_object().unwrap() {
            body[key] = value.clone();
        }
        let error = call(service, "POST", "/api/automations", body)
            .await
            .unwrap_err();
        assert!(
            format!("{error:#}").contains(expected),
            "{expected}: {error:#}"
        );
    }
    let preview = call(
        service,
        "POST",
        "/api/automations/preview",
        json!({"schedule": {"kind": "weekdays", "time": "09:00"}, "timezone": "utc"}),
    )
    .await
    .unwrap();
    assert_eq!(preview["ok"], true);
    assert_eq!(preview["description"], "Weekdays at 09:00 UTC");
    assert_eq!(preview["next"].as_array().unwrap().len(), 3);
    let bad = call(
        service,
        "POST",
        "/api/automations/preview",
        json!({"schedule": {"kind": "cron", "expr": "nope"}}),
    )
    .await
    .unwrap();
    assert_eq!(bad["ok"], false);

    Config::update(&fixture.paths, |cfg| {
        cfg.trusted_workspaces.clear();
        Ok(())
    })
    .unwrap();
    let error = call(
        service,
        "POST",
        "/api/automations",
        automation("Untrusted", "main"),
    )
    .await
    .unwrap_err();
    assert!(error.to_string().contains("Trust"), "{error}");
}

#[test]
fn runs_interrupted_by_a_restart_are_recovered() {
    let root = tempfile::tempdir().unwrap();
    let db = root.path().join("state.sqlite");
    let store = Store::open(&db).unwrap();
    let run = AutomationRun {
        id: "r1".into(),
        automation_id: "a1".into(),
        status: "running".into(),
        trigger: "schedule".into(),
        started_at: 10.0,
        ..Default::default()
    };
    store.save_automation_run(&run).unwrap();
    drop(store);
    let store = Store::open(&db).unwrap();
    assert_eq!(store.recover_automation_runs().unwrap(), 1);
    let recovered = store.automation_run("r1").unwrap();
    assert_eq!(recovered.status, "interrupted");
    assert!(recovered.finished_at.is_some());
    assert!(recovered
        .detail
        .contains("stopped before this run finished"));
    assert_eq!(store.recover_automation_runs().unwrap(), 0);
}
